### ISSUE TITLE
Language handing in Book model

### ISSUE THREAD

**Issue by aristippe** [aristippe/pathagar#3]
The language model is currently set with a code max length of 5 characters. I forgot what RFC5646 specifies, though there is quite a variation in how languages are specified in different ePubs. How to handle it? Require spec, allow for variations but store it per spec in the DB, or store whatever string specified in ePub?

A place for admins, or other future user classes such as superuser, moderator, etc, to change language strings without going into Django admin would also be a good future addition.

**Comment by sinergatis** [comment]
For reference, an existing Python library that seems to handle RFC5646.
https://github.com/OnroerendErfgoed/language-tags

Ideally we should check the EPUB2 and other applicable standards to find out if RFC5646 covers those as well. Some degree of "fuziness" would probably be indeed a good idea, as it seems that epubs on the wild might not fully adhere to the standards, sadly.

**Comment by aristippe** [comment]
One thing I noticed was that while editing an ePub, not working btw, all languages show up, so if there's multiple for one, e.g., fre, fr, fr_fr, etc, they're all included. Perhaps there should be some intermediate lookup of possible languages specified in an ePub, but for the web UI, a separate table to show only unique values?

**Comment by sinergatis** [comment]
> One thing I noticed was that while editing an ePub, not working btw

Noticed, and something that needs looking into - 2fdfe58a26b26eb7930f59d503fb652517316c86 (not yet merged, as it is included with the round of django upgrades) tries to take care of this, although needs further work.

> all languages show up, so if there's multiple for one, e.g., fre, fr, fr_fr, etc, they're all included. Perhaps there should be some intermediate lookup of possible languages specified in an ePub, but for the web UI, a separate table to show only unique values?

Sounds like a good idea, although it might be a side effect of the current status of the `Language` handling and the current fixtures. We might temporarily want to include the "short" string on the Language description (so they show up as `French (fr_fr)`), but some way of filtering them and just showing the "main" variation to the user sounds very reasonable.

**Comment by sinergatis** [comment]
A request on this one, since you are probably the one with the most "representative" library: could you do something like:

``` bash
$  python manage.py addepub /your/library/path | grep 'Language'
```

whenever you have the chance, and pass me the results somehow? It would be a nice benchmark for finding out how many epubs actually deviate from the standard, and have a more solid ground for deciding on this issue.

EDIT: a grep on just `Language` perhaps makes more sense, as having info about epubs where no language was found can also be useful!

**Comment by aristippe** [comment]
Here ya go. :)

[lang.txt](https://github.com/aristippe/pathagar/files/128504/lang.txt)

Run on a set of about 3800 books. I have a larger set on a server I can later try, though I'll have to manage a test deploy with Apache/mod-wsgi or Nginx. Even on my own machine I haven't quite gotten that yet.

And thank you for fixing the book edit!

**Comment by sinergatis** [comment]
Amazing, thanks a lot! At a first glance, it seems that, ignoring uppercase/lowercase, a very minor number of books differ from the standard, which is ... great news! I'll check it more carefully later on, but I'm positively surprised :tada:

**Comment by sinergatis** [comment]
I'm working on a solution that enforces RFC5646 for Epub3 and RFC3066 for Epub2, and also handles ISO-6639/2 bibliographic codes (in particular for the `fre` tag - there seems to be some slight problems with ~20 codes where a ISO-6639/2 subset allows for slightly different versions of the tag (even if `fra` is supposed to be the right one)). A summary of the results so far:

| Count | epub tag | Old label | New code | New description |
| --- | --- | --- | --- | --- |
| 28 | de | German | de | German |
| 4 | el | Modern Greek (1453-) | el | Modern Greek (1453-) |
| 1172 | en | English | en | English |
| 57 | En |  | en | English |
| 53 | EN |  | en | English |
| 65 | eng |  | en | English |
| 3 | en-gb |  | en-gb | English, United Kingdom |
| 43 | en-GB |  | en-gb | English, United Kingdom |
| 4 | English |  |  |  |
| 1 | en-UK |  |  |  |
| 23 | en-us |  | en-us | English, United States |
| 1 | en_US |  |  |  |
| 326 | en-US |  | en-us | English, United States |
| 1 | eo | Esperanto | eo | Esperanto |
| 12 | es | Spanish | es | Spanish, Castilian |
| 3 | ES |  | es | Spanish, Castilian |
| 1698 | fr | French | fr | French |
| 15 | fra |  | fr | French |
| 2 | fr-CA |  | fr-ca | French, Canada |
| 160 | fre |  | fr | French |
| 51 | fr-FR |  | fr-fr | French, France |
| 81 | it | Italian | it | Italian |
| 3 | IT |  | it | Italian |
| 7 | it-IT |  | it-it | Italian, Italy |
| 1 | nl | Dutch | nl | Dutch, Flemish |
| 1 | pt | Portuguese | pt | Portuguese |
| 1 | ru | Russian | ru | Russian |
| 1 | Spanish |  |  |  |
| 3 | UND |  | und | Undetermined |
| 1 | zh | Chinese | zh | Chinese |
| 1 | zh-CN |  | zh-cn | Chinese, China |

I still need to polish some things, and involves a couple of dependencies, but it's looking promising (only ~8 epubs on the previous `lang.txt` sample result in invalid tags, compared to ~700 on the previous version of the code!).

**Comment by sinergatis** [comment]
Closed by #51!
